### PR TITLE
Turn on iOS wireless debugging, connect to remote observatory

### PR DESCRIPTION
**Pull request by jmagman**
## Description
Use new `--network` libimobiledevice flags.

`flutter run`, `flutter attach --host-vmservice-port`, `flutter install`, `flutter screenshot` work.

## Related issues
https://github.com/flutter/flutter/issues/15072

## Steps to test
1. Attach a phone to your Mac.  This works best when both are on the same Wi-Fi network (unrelated to Flutter).
1. Make sure the device has a passcode set.
1. Open Xcode > Window > Devices and Simulators > Choose your phone, then check Connect via Network
Screen Shot 2020-06-26 at 7 40 35 PM
(If you don't get that little network icon next to the device name it probably didn't work, regardless of checkbox state.)
1. Unplug your device from USB.

## Tests
- mac_test.dart
- xcode_test.dart

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

**Comment by jmagman** [comment]
\cc @zanderso @DanTup @dnfield @jonahwilliams Any insight on this one?

**Comment by jmagman** [comment]
@anirudhb Were you ever actually able to get https://github.com/flutter/flutter/pull/58185 working and attaching to the observatory? (Thank you again for that contribution! Fortunately the newest version of `iproxy` handles the IP lookup for us, so we don't need to do any stdout parsing).  I also got "Connection refused" issues when I ran your PR, though it may have been because the IP wasn't parsed correctly on my machine, and even when the host and target were on the same network I still get the `Non-authoritative` output (maybe a hint?).

If you did have it working on your PR, would you mind testing this PR with the same setup?  Did you have to set up anything special to get it working?

**Comment by anirudhb** [comment]
I didn’t try anything yet but glancing at the observatory logs it seems you forgot to pass `—observatory-host 0.0.0.0` to the app arguments.

**Comment by jmagman** [comment]
> I didn’t try anything yet but glancing at the observatory logs it seems you forgot to pass `—observatory-host 0.0.0.0` to the app arguments.

@anirudhb Good eye, that was totally it.  Thank you so much!

**Comment by anirudhb** [comment]
Regarding the performance of this, it seems that install always copies all files to the device (including Flutter.framework) which is unreasonably large and slow. Have you looked into making Flutter's build work with app deltas? (This doesn't work currently since the build script seems to be touching Flutter.framework, forcing it to be copied.)

**Comment by jmagman** [comment]
> Regarding the performance of this, it seems that install always copies all files to the device (including Flutter.framework) which is unreasonably large and slow. Have you looked into making Flutter's build work with app deltas? (This doesn't work currently since the build script seems to be touching Flutter.framework, forcing it to be copied.)

The version of ios-deploy we're embedding doesn't support [app-deltas](https://github.com/ios-control/ios-deploy/pull/434) yet (we're back on `1.10.0-beta.3`), but we'll need to upgrade soon to support Xcode 12 (see https://github.com/flutter/flutter/issues/60072#issuecomment-655762513).  We can explore that separately since it'll help non-wireless install times as well.

**Comment by anirudhb** [comment]
fwiw I did try the new ios-deploy with app deltas, but it seems that Flutter builds the app in such a way that app deltas copies the entire app again (including Flutter.framework, which shouldn't've even changed.)

**Comment by jmagman** [comment]
> fwiw I did try the new ios-deploy with app deltas, but it seems that Flutter builds the app in such a way that app deltas copies the entire app again (including Flutter.framework, which shouldn't've even changed.)
> 
> The app install time is extremely noticeable over wireless, whereas over USB it's not as significant.

That doesn't surprise me, we scorch earth that framework on every build because people kept submitting the debug version to the App Store, which gets at best gets a rejection and at worst causes a 100% crash on launch.  I've hopefully plugged enough holes with that logic that maybe we can be less aggressive, but we'll have to investigate further.  I'm comfortable turning this feature on even if it's painfully slow, and we can improve later.

**Comment by anirudhb** [comment]
I did some investigation on this, and the gist of it is that Xcode only uses modification time to determine whether to copy a file or not. Some things in the build that change mtime include code signing and framework thinning. Cocoapods additionally does this as well (this could be an upstream issue). After modifying the scripts to preserve mtime across code-sign/thin, Xcode was successfully able to use app deltas!

A proper implementation of this would probably involve hash comparisons, since Xcode doesn't do this for some reason when computing app deltas.

**Comment by jmagman** [comment]
I removed the longer timeouts in favor of a new flag https://github.com/flutter/flutter/pull/64834.  With both this change and that one, it's working on iOS 13 and Xcode 11.  Unfortunately I'm hitting [debug attach issues](https://github.com/ios-control/ios-deploy/pull/471#issuecomment-683190276) on Xcode 12/iOS 14.

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 1]
```diff
@@ -364,6 +355,8 @@ class IOSDevice extends Device {
       '--enable-service-port-fallback',
       '--disable-service-auth-codes',
       '--observatory-port=$assumedObservatoryPort',
+      if (interfaceType == IOSDeviceInterface.network)
+        '--observatory-host=${ipv6 ? '::/0' : '0.0.0.0'}',
```
Comment (xster) [inline packages/flutter_tools/lib/src/ios/devices.dart:359]: add comment for reader

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 2]
```diff
@@ -762,10 +756,12 @@ class IOSDevicePortForwarder extends DevicePortForwarder {
     @required String id,
     @required IProxy iproxy,
     @required OperatingSystemUtils operatingSystemUtils,
+    @required IOSDeviceInterface interfaceType,
```
Comment (xster) [inline packages/flutter_tools/lib/src/ios/devices.dart:759]: ultra nit / feel free to ignore: IOSDeviceConnectionInterface? Just interface is a bit vague.

**Review by xster — COMMENTED** [review]
LGTM

**Review by christopherfujino — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 1 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/ios/devices.dart:360]: Also see https://github.com/flutter/engine/pull/9485.

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 2 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/ios/devices.dart]: Done.

**Comment by jmagman** [comment]
Not working on iOS 14 again...

**Comment by christopherfujino** [comment]
I guess re-requesting a review doesn't remove an approval, as far as GitHub is concerned?

**Review by christopherfujino — CHANGES_REQUESTED** [review]
Just trying to clear out my previous approval...in case the mergebot goes crazy again and tries to merge it.

**Comment by christopherfujino** [comment]
> Just trying to clear out my previous approval...in case the mergebot goes crazy again and tries to merge it.

I look like a genius now :)

**Comment by jmagman** [comment]
This is well out of date, closing.

**Pull request closed without merging**